### ISSUE TITLE
MAVLink battery information

### ISSUE THREAD

**Issue by LorenzMeier** [mavlink/mavlink#988]
Questions the message should answer:

 * Charge level
 * Health
 * How long it will last
 * Age / max charge level / capacity
 * Battery ID
 * Any fault indications

```xml
    <message id="334" name="BATTERY_INFO">
      <description>Battery information to communicate once per power cycle.</description>
      <field type="uint32_t" name="serial_number">Battery serial number, UINT32_MAX: battery does not have a serial number</field>
      <field type="uint32_t" name="capacity" units="mAh">Actual capacity of the battery measured by the fuel gauge, UINT32_MAX: battery does not report actual capacity</field>
      <field type="uint16_t" name="cycle_count">Number of charge-discharge cycles reported by the fuel gauge, UINT16_MAX: battery does not report cycle count</field>
    </message>
```

**Comment by hamishwillee** [comment]
There is an [old spec here](http://sbs-forum.org/specs/sbdat110.pdf) that might seed additional information we could include. In particular I suspect it would be useful to be able to remotely monitor the progress of a battery as it is being charged (i.e. if it is an internal battery on a remotely managed charging system).

The messages talked about in this spec are:
```
ManufacturerAccess() (0x00) 13
5.1.2. RemainingCapacityAlarm() (0x01) 13
5.1.3. RemainingTimeAlarm() (0x02) 13
5.1.4. BatteryMode() (0x03) 14
5.1.5. AtRate() (0x04) 21
5.1.6. AtRateTimeToFull() (0x05) 22
5.1.7. AtRateTimeToEmpty() (0x06) 22
5.1.8. AtRateOK() (0x07) 23
5.1.9. Temperature() (0x08) 23
5.1.10. Voltage() (0x09) 24
5.1.11. Current() (0x0a) 24
5.1.12. AverageCurrent() (0x0b) 24
5.1.13. MaxError() (0x0c) 25
5.1.14. RelativeStateOfCharge() (0x0d) 25
 RemainingCapacity() (0x0f) 26
5.1.17. FullChargeCapacity() (0x10) 26
5.1.18. RunTimeToEmpty() (0x11) 27
5.1.19. AverageTimeToEmpty() (0x12) 27
5.1.20. AverageTimeToFull() (0x13) 27
5.1.21. BatteryStatus() (0x16) 28
5.1.22. CycleCount() (0x17) 32
5.1.23. DesignCapacity() (0x18) 32
5.1.24. DesignVoltage() (0x19) 32
5.1.25. SpecificationInfo() (0x1a) 33
5.1.26. ManufactureDate() (0x1b) 34
5.1.27. SerialNumber() (0x1c) 34
5.1.28. ManufacturerName() (0x20) 34
5.1.29. DeviceName() (0x21) 34
5.1.30. DeviceChemistry() (0x22) 34
5.1.31. ManufacturerData() (0x23) 35
5.2. Smart Battery or SMBus Host to Smart Battery Charger Messages 36
5.2.1. ChargingCurrent() (0x14) 36
5.2.2. ChargingVoltage() (0x15) 37
5.3. Smart Battery Charger or SMBus Host to Smart Battery Messages 38
5.3.1. ChargingCurrent() (0x14) 38
5.3.2. ChargingVoltage() (0x15) 39
5.4. Smart Battery Critical Messages 40
5.4.1. AlarmWarning() (0x16)
```

**Comment by AlexKlimaj** [comment]
If we are going to create a brand new smart battery message, here is what I would include as a starting point. 

Pack voltage
Current (Positive if discharging like the existing message, negative if charging)
Temperature
RelativeStateOfCharge
FullChargeCapacity
State of Health (%)
RunTimeToEmpty
CycleCount
SpecificationInfo
SerialNumber

For @dagar 's point, I agree the driver should be smart enough to determine if the battery has failed and communicate the failure mode. Maybe we create one status field with enum's of the different states and failures. 
And maybe one more enum field to communicate the last detected fault but not failure. ie short circuit, overload, over temperature, under temperature, over voltage, etc. These events could have occurred when the battery was not connected to the autopilot. 

These below I think could be included or not.

DeviceName
DeviceChemistry
ManufacturerName
ManufactureDate
ChargingVoltage
ChargingCurrent
BatteryStatus
AverageCurrent

**Comment by hamishwillee** [comment]
@LorenzMeier @MaEtUgR et al, are we at the point where a proposed "smart battery message could be constructed? How/who to progress to the next step?

**Comment by MaEtUgR** [comment]
We could do a message, I currently just don't have any smart battery to integrate and test...
High level I vote for the following in priority order:
- Charge level (persent/ratio)
- Capacity when full (with degradation accounted for)
- Current (through the cells, all loads)
- Fault indications
(containing volatile flags like voltage spikes and persistent flags indicating health e.g. deep discharge)
- Cell Voltages (telemetry/checks for precautious reactions)
- Unique ID / Serial

**Comment by hamishwillee** [comment]
@AlexKlimaj @MaEtUgR @LorenzMeier  I've marked this for tomorrow's MAVLink devcall. Might be great if you could come along to discuss? It is 6pm AEST (Melbourne time).

Combining your thoughts give us the below. Would be good if you could respond with perhaps a few reasons why you exclude the options that each other have included. For instance, it seems to me that specific fault information would be useful, also estimate of remaining fly time. Who is the consumer of the message - GCS? What can the user do with info/what do they need? Why don't you need to know the serial number - isn't that useful for identifying one of many batteries?

#### Included fields
@MaEtUgR + @AlexKlimaj 

- Charge level (%/ratio)
- Capacity when full (with degradation accounted for)
- Current
  - through the cells, all loads
  - Positive if discharging like the existing message, negative if charging

@MaEtUgR only
- Fault indications
(containing volatile flags like voltage spikes and persistent flags indicating health e.g. deep discharge)
- Cell Voltages (telemetry/checks for precautious reactions)

@AlexKlimaj 
- Pack voltage
- Temperature
- RelativeStateOfCharge
- State of Health (%)
- RunTimeToEmpty
- CycleCount
- SpecificationInfo
- SerialNumber

#### Optional/ambivalent
@AlexKlimaj  + @MaEtUgR 
- DeviceChemistry
- ManufacturerName
- ManufactureDate

@AlexKlimaj 
- DeviceName
- ChargingVoltage
- ChargingCurrent
- BatteryStatus
- AverageCurrent

@MaEtUgR 
- Number of cells
- Nominal-voltage per cell
- Serial No

**Comment by MaEtUgR** [comment]
@hamishwillee Thanks for the summary.
- Pack voltage
This information is in the mandatory "Number of cells", "Nominal-voltage per cell" for nominal and "Cell Voltages" for the the momentary voltage under load.

- Temperature
Agree, nice to have if the battery has a sensor.

- RelativeStateOfCharge
For me that's the "Charge level". What does it help if you know how charged it would be if it was still new.

- State of Health (%)
How should this be defined? I think it's to general and doesn't help to draw conclusions in the end. I concentrated to have the most important infomation that lets everyone draw his health conclusions based on the latest state of art and his requirements.

- RunTimeToEmpty
It's "Capacity when full" * "Charge level" / "Current". The difficulty is to filter it adequately to account for the expected average load until the battery is empty. And this task should not be done by the battery itself because it lacks additional information that the flight control unit might have like e.g. vehicle model, best case, worst case, pilot habits from vehicle average current over all flights, ...

- CycleCount
Forgot that one. Totally makes sense to add this to the age specification together with ManufactureDate. It says a lot about the battery.

- SpecificationInfo
Is that C rating charge rating and so on? That would also be nice to know statically when it gets plugged.

- SerialNumber
Same as "Serial No"

If I may edit the summary in the last comment by @hamishwillee , the history stays visible (since github added the feature) and I think it's more clear to read than when I copy everything again.

**Comment by hamishwillee** [comment]
> @hamishwillee , the history stays visible (since github added the feature) and I think it's more clear to read. I'm not allowed to edit :)

Go for it @MaEtUgR  - I'm here to facilitate, not stop you doing the right thing. If you can't edit for whatever reason, send me the text and I'll copy paste tomorrow. 

One thing though, I agree that the below should not come from the battery. I was (perhaps foolishly) assuming most of this info would come via the flight controller to consumers. 

>  And this task should not be done by the battery itself because it lacks additional information that the flight control unit might have like e.g. vehicle model, best case, worst case, pilot habits from vehicle average current over all flights, ...

**Comment by apache405** [comment]
Overall, the message contents look pretty good to me and I am happy to see these.

I have a few items/thoughts that I'd like to make from the point of view of someone who has built and maintained several smart battery designs.

Regarding the comments that @hamishwillee and @MaEtUgR made about the estimated remaining runtime, the battery manager ICs that most people get from TI, etc. are pretty dumb when it comes to calculating runtime to empty (you have a limited set of options and they all are bad/incomplete), so given today's BMS/Fuel gauge options I think the approach of the flight controller making the runtime to empty estimates is reasonable.  However, if the BMS feature landscape changes, I think this item should be revised. 

Another thing that I would greatly appreciate is the ability to handle real-time streaming of BMS data back to a ground station via MAVLINK (manufacturer info blocks mostly).  For example, TI has tools that let me tune the performance of the BMS, but I am only able to do this in real time if I strap their development interface box to the aircraft and tether things back to a laptop over a long USB cable.  I would greatly prefer to have the ability to do this over the telemetry link.

**Comment by Naterater** [comment]
The number one item in my mind is live cell voltages. A single cell overdrawn can cause the whole pack to ignite, and rarely do all cells have the exact same lifespan. Post-flight voltage recover and balancing only do so much. Monitoring cell voltages and differentials (not just nominal cell voltage) is critical to me.

Although not mentioned by anyone here, **battery weight** could be useful for dynamic delivery vehicles that might have different tunes or parameters depending on the vehicle's total weight. Battery weight is often a critical choice, and can often be modified depending on payload. The vehicles I have use different batteries when different payloads are installed, and it would be beneficial to automatically prevent overweight takeoffs when certain equipment is present.

**Comment by hamishwillee** [comment]
@apache405 Thanks very much for the feedback! BMS is battery management system?

We discussed this in the MAVLink dev call last night. You're right - there are multiple use cases this message should address:
- MAVlink smart battery sending messages to autopilot and/or GCS
- Autopilot sending battery messages to GCS
- Bridging messages between UAVCAN/SMBUS and MAVLink.

Specifically for the case of "fuel gauge" options, we thought that there is no reason that the battery might not send the message with whatever info it has to the autopilot, and then the autopilot send the same message type to a GCS, adding its estimate of fuel gauge. We don't mandate any field "must" be filled in. 

What we do want to do though is make sure that the set of fields we define allow the above use cases for at least the most important information. Do you think the set in https://github.com/mavlink/mavlink/issues/988#issuecomment-446017379 makes sense for the above SMBUS/UAVCAN cases, and your own needs?

**Comment by hamishwillee** [comment]
All, I have edited https://github.com/mavlink/mavlink/issues/988#issuecomment-446017379 to redefine the list as proposed in/out fields with additional fields based on further feedback from @MaEtUgR 
- @Naterater Thanks! The above list includes battery weight (cell voltages already agreed). Of course if we get strong disagreement things might come out. For example, I wonder if a system is able to "know" its battery weight programmatically? How often would this be needed?
- @apache405 I added the *Estimated Run Time to Empty*. This can be provided (or not) by any system. It is up to the recipient to decide what version it wants to use. I think it is useful for a GCS or even better an FrSky transmitter to be able to have this guess.
- I added battery index on assumption a vehicle might have more than one. Is that reasonable?

So this is now a proposed set. Can we all look again and let me know if you are happy with the list as it stands? If you had to add another item to the top list, what would it be?

**Comment by AlexKlimaj** [comment]
Hey everyone, I apologize for not joining the discussion earlier. I've been away on business travel. I'm glad this has become a topic of discussion again. 

- @hamishwillee Regarding the **battery weight**, I think it is useful. You could use the battery weight information to change PID tuning depending on what weight of battery is used. 
- **State of Health (%)** - I think this is still useful to have. It is a measure of actual capacity/nominal capacity. It's a quick measure of the wear level on a battery. This, or have a nominal capacity field so a GCS could calculate it. 
- **AverageCurrent** - on the BQ40Z50, it is a moving average of 15 seconds. May be useful or not. 
- **Estimated Run Time to Empty** - @apache405 I agree. Our BQ40Z50 is pretty dumb with its calculation of this. But with a more advanced fuel gauge in the future, this could be very useful for autonomous flight with a companion computer monitoring this and optimizing the flight for maximizing runtime. But I guess this could also be done by a companion computer calculating power from voltage and current.

**Comment by hamishwillee** [comment]
@pavel-kirienko Thank you. Is there anything else missing in https://github.com/mavlink/mavlink/issues/988#issuecomment-446017379 that you think might be required for bridging smart battery information from UAVCAN?

@MaEtUgR Can you please comment/respond on @pavel-kirienko messages above - ie whether you see need to change the list of what is in/out and why. 

@AlexKlimaj Thanks for comments above. Everyone is busy - appreciate you rejoining. 
- Re battery weight, I can see that it might be useful, but is it *feasible*? ie what system would populate this information, and where would it get the information?
- I tend to think `State of Health` is a "nice to have" to allow easy display on GCS - but you might better have this as just states red/orange/green (broken/risky/good) rather than percentage. 
  - As long as the info required to calculate those states is present I wouldn't see it as essential. 
  - What should a GCS do if it gets both the raw data and the stateofhealth?
- I'm going to assume AverageCurrent  is not needed unless someone stands up and says they need it (and why).
- Yes, fuel gauge is potentially useful. It is in the top list - but is not "mandatory" to populate. Further comment on that in https://github.com/mavlink/mavlink/issues/988#issuecomment-446832744

**Comment by hamishwillee** [comment]
Do we need a battery index if we have a serial number?

**Comment by WickedShell** [comment]
> Do we need a battery index if we have a serial number?

If you have a function field then maybe not, otherwise you don't have data to correlate what that is actually powering on the vehicle.

**Comment by hamishwillee** [comment]
>> Do we need a battery index if we have a serial number?

> If you have a function field then maybe not, otherwise you don't have data to correlate what that is actually powering on the vehicle.

@WickedShell What is a function field? I *guess* you are saying that the index is to map to some field on an autopilot - e.g. battery one, two, three and you need to know which in order to map to other power information on the GCS?

**Comment by Naterater** [comment]
> @hamishwillee @MaEtUgR Pack voltage is also a function of the battery configuration: cells can be connected in series and in parallel. Unless the configuration is known, the pack voltage can't be computed from the number of cells and their voltages. It is quite possible though that I am just missing some context here because I just saw this conversation for the first time.

Unless I'm mislead, cells that are in parallel with each other should always be permanently and physically cross-balanced. I've never seen a pack that wasn't, therefore the sum of the cells in series == the total battery voltage. Always.

**Comment by AlexKlimaj** [comment]
@Naterater You are correct. Any battery that has cells configured in parallel, will essentially be 'one' cell in series to the fuel gauge. So you could have a 4s3p smart battery, but the fuel gauge thinks it only has 4 cells in series. 

If we are assuming people in the future will be using more than one smart battery, they probably wont end up having all the cells cross-balanced between the two packs. So they can just be treated as two completely separate batteries. 

I guess what I'm trying to get at, is that we really only care about number of cells in series. Then we also care about the total number of smart battery packs, which would be a whole separate instance of the message.

**Comment by apache405** [comment]
@hamishwillee  
> BMS is battery management system?
Correct.

> Do we need a battery index if we have a serial number?
I think we'll be forced into serial numbers for where-used/when-used traceability.  The index thing might be nice for mapping which port something is connected to, but I'd trade it if in favor of serial numbers and other message items if things get tight on the message sizing.

@AlexKlimaj  > * **AverageCurrent** - on the BQ40Z50, it is a moving average of 15 seconds. May be useful or not.
Yeah, can be useful; but there are quite a number of data items and status items I'd rather see go in first. 

IIRC, almost all of the analog items reported from a 40z50 and 40z60 are already filtered in both analog and digital domains before they reach a place we can read them.  The ADC grabs a sample at 4Hz and the SMBus registers update at 1Hz.  The value in the register is the average of the last 4 samples.

> So they can just be treated as two completely separate batteries. 
In general, supporting multiple batteries is very important, especially on larger vehicles or on vehicles that need multiple types of battery (example: quad-planes with a high-c pack for take-off and a cruise pack for post-transition flight)

**Comment by WickedShell** [comment]
> @WickedShell What is a function field? I guess you are saying that the index is to map to some field on an autopilot - e.g. battery one, two, three and you need to know which in order to map to other power information on the GCS?

Your GCS gets 2 battery messages, with hopefully different (but not necessarily) serial numbers. How do you know which one is powering your forward motor and which is powering your payload? You either need a function field to convey this (which has a serious flexibility problem as the enum has to cover all the uses you want, and still doesn't allow you to use 2 batteries in parallel for the same function), or a unique number on the aircraft that you map via a sidechannel to a meaningful value for the user (IE extract from param data, a file on your GCS.

**Comment by pavel-kirienko** [comment]
> the sum of the cells in series == the total battery voltage.

> we really only care about number of cells in series.

@AlexKlimaj @Naterater Makes sense. This logic probably should be reflected in the name of relevant fields and/or in the documentation, to make it explicit that a "cell", from the standpoint of the protocol, is not a real physical cell, but rather a set of cells connected in parallel. This is important to clarify because it differs from the definition of cell given in the Smart Battery Data Specification:

> Cell: The cell is the smallest unit in a battery. Most batteries consist of several cells connected in series, parallel, or series-parallel combination.

@hamishwillee I don't think there's anything missing here.

**Comment by hamishwillee** [comment]
All - @AlexKlimaj @pavel-kirienko @WickedShell @apache405 @Naterater @LorenzMeier 

Thanks everyone for your feedback/info. I think we're at a good list (not, now includes battery weight and index). Can all of you please vote thumbs up/down in https://github.com/mavlink/mavlink/issues/988#issuecomment-446017379 to confirm no issues with creating PR with that set. If you do thumbs down, propose change please.

Anyone want to volunteer for that PR creation (otherwise I will do it).

**Comment by hamishwillee** [comment]
Hi All, 

Below is first shot at message definition "for discussion". There are some open questions/issues at end. Remember, I know very little about smart batteries - I'm being led by you!

In general, consider 
- "are the units and storage types correct"?
- Are names and descriptions sensible/unambiguous

@MaEtUgR First the bitmask enum for battery faults:
- What other faults are there?
- What should recipient do in each case? i.e.:
  - When might voltage spikes be reported. This is transitory - what would a recipient do. Ditto for deep discharge warning?

```xml
    <enum name="MAV_SMART_BATTERY_FAULT">
      <description>Smart battery supply status/fault flags (bitmask) for health indication.</description>
      <entry value="0" name="MAV_SMART_BATTERY_FAULT_OK">
        <description>Battery has no fault indications.</description>
      </entry>
      <entry value="1" name="MAV_SMART_BATTERY_FAULT_DEEP_DISCHARGE">
        <description>Battery has deep discharged.</description>
      </entry>
      <entry value="2" name="MAV_SMART_BATTERY_FAULT_SPIKES">
        <description>Voltage spikes.</description>
      </entry>
      <entry value="4" name="MAV_SMART_BATTERY_FAULT_?">
        <description>?</description>
      </entry>
    </enum>
```

This is the message:
```xml
    <message id="365" name="SMART_BATTERY_STATUS">
      <description>Smart Battery information.</description>
      <field type="uint8_t" name="id">Battery ID</field>
      <field type="int8_t" name="capacity_remaining" units="%">Remaining battery energy. Values: [0-100], NaN: field not provided.</field>
      <field type="int32_t" name="capacity_full" units="mAh">Capacity when full (accounting for battery degradation), NaN: field not provided.</field>
      <field type="int16_t" name="current" units="cA">Battery current (through all cells/loads). Positive if discharging, negative if charging. NaN: field not provided.</field>
      <field type="int16_t" name="temperature" units="cdegC">Battery temperature. NaN: field not provided.</field>
      <field type="uint16_t[10]" name="voltages" units="mV">Individual cell voltages. NaN: Invalid cells (i.e. index values above the battery cell count).</field>
      <field type="uint16_t" name="cycle_count">Charge/discharge cycle count. NaN: field not provided.</field>
      <field type="int32_t" name="serial_number">Serial number. NaN: field not provided.</field>
      <field type="int32_t" name="fault_bitmask" display="bitmask" enum="MAV_SMART_BATTERY_FAULT">Fault/health indications. NaN: field not provided.</field>
      <field type="uint16_t" name="weight" units="g">Battery weight. NaN: field not provided.</field>
      <field type="int32_t" name="time_remaining" units="s">Estimated remaining battery time. NaN: field not provided.</field>
      <field type="char[50]" name="device_name">Static device name. Encode as manufacturer and product names separated using an underscore.</field>
    </message>
```
1. What should units be for battery weight. I assumed grams, and an uint16, giving us max 65kg for smart batter. Reasonable? Need to use finer resolution (mg) and/or bigger field - uint32?
   - Note, there are no mass units in schema, so that will also need to be updated [here](https://github.com/ArduPilot/pymavlink/blob/master/generator/mavschema.xsd)
1. @MaEtUgR I looked into `SpecificationInfo`. This is a version number for the current specification in smartbus - ie they carry version info in the interface. We don't. You stated "Is that C rating charge rating and so on? That would also be nice to know statically when it gets plugged."
   - Now we know it isn't, can you add example of the field(s) you'd like.
1. Is "10" enough cells?
1.  How many chars do we need for the device name? This is at end so it will be cropped "now" but in future we might have an extension, so we do want a reasonable value. 
1. Is NaN the *right* think to do for "not supported"?
1. How would I capture @pavel-kirienko point https://github.com/mavlink/mavlink/issues/988#issuecomment-447259221 about series vs parallel cells above?

**Comment by Naterater** [comment]
> Is "10" enough cells?

No - I would go at least to 12. 6s batteries in series is not too difficult or even uncommon. 16 should cover everything in the future as well in my opinion.

For faults - "MAV_SMART_BATTERY_FAULT_DELTA" (one cell failing but not the rest) could be an idea.

I think weight is good. 65kg should get us far.

Semi-on topic - could a "smart fuel tank" also be able to use this protocol? I'm thinking of a fuel gauge or a flow meter. There are a lot of parallels here, and weight would transfer as well. Maybe we could phrase some things to be intentionally cross-compatible.

Another thing that might be useful is a minimum voltage per cell field. Li-ion cells can often go much lower than li-po cells. Configuring voltage failsafes based on this type of parameter might be very useful - even more-so than weight.

**Comment by hamishwillee** [comment]
@Naterater Thanks. 
- I updated the battery voltages to 16 in post above.
- I added MAV_SMART_BATTERY_FAULT_DELTA - why that name? (ie why not MAV_SMART_BATTERY_FAULT_SINGLE_CELL_FAIL). 
- Re minimum voltage per cell - is it worth adding another 16 `uint16_t` or can these all be assumed to be the same voltage? @AlexKlimaj @MaEtUgR @apache405 What do you think on this one?

> Semi-on topic - could a "smart fuel tank" also be able to use this protocol? I'm thinking of a fuel gauge or a flow meter. There are a lot of parallels here, and weight would transfer as well. Maybe we could phrase some things to be intentionally cross-compatible.

Would be awesome, and I'm happy to take suggestions of a proposed message that does this well. But I'm skeptical. The problem is the units that won't apply, and also some of the fields (e.g. charge cycle). I worry we'd end up with an inefficient message that was a poor compromise for either case.

**Comment by MaEtUgR** [comment]
I think it's important to **distinguish static information** like ID, chemistry, weight, ... and dynamic information like voltage and current under load. Because like we can see here, we'll end up with a superset of information and if we stream static information with a reasonable rate it's just a huge waste of communication resources.

> are pretty dumb when it comes to calculating runtime
> on the BQ40Z50, it is a moving average of 15 seconds
In my opinion these roughly preprocessed values are not useful if you anyways have a flight control battery estimation which can be fed with more information and tuned to specific needs. Also if you're speaking about advanced monitoring algorithms it should not be the task of the smart battery because it should be independent of the vehicle and not unnecessarily process information.

> The number one item in my mind is live cell voltages.
I thought the same 👍 

> battery weight could be useful for dynamic delivery vehicles that might have different tunes or parameters depending on the vehicle's total weight

I honestly didn't think about this but I it's a good idea for one for a static field which only needs to be communicated once per battery. Even the simplest models usually include the weight.

> State of Health ... or have a nominal capacity field so a GCS could calculate it.
That's exactly what I would do. In the static information have the nominal (manufacturer specification) and available (possibly degraded) full capacity and in the dynamic information the left capacity with the current charge. This way it's unambiguous.

**Comment by Naterater** [comment]
> Re minimum voltage per cell - is it worth adding another 16 uint16_t or can these all be assumed to be the same voltage?

Minimum voltage per cell should be the same across all cells - a single static value (ex. 3.60v) would be acceptable. Mixed chemistry batteries (individual low voltage values per cell) is something that I don't think is practical or currently available.

> I think it's important to distinguish static information like ID, chemistry, weight, ... and dynamic information like voltage and current under load. Because like we can see here, we'll end up with a superset of information and if we stream static information with a reasonable rate it's just a huge waste of communication resources.

Sounds reasonable. It sounds to me like the static/dynamic list would look like the following

Static:
Battery ID
Capacity Full
Cycle Count
Serial Number
Device Name
Weight
Minimum Voltage

Dynamic:
Capacity Remaining
Current
Voltages
Fault Bitmask
Time Remaining

Regarding supporting a fuel tank - I agree that this should be focused on batteries. If it doesn't work, disregard. It's not as important as smart batteries. I just know piccolo uses fuel weight calculations, and this would be a step toward that. Since live weight can be calculated by capacity remaining, it belongs as a static value to save communication resources.

For what it's worth, I think the time remaining message is redundant to the capacity remaining. I think it should be removed. The time remaining should be something calculated externally. Capacity (as a percentage) should be mapped to reflect the true energy capacity of the battery at its current discharge rate so that time calculations can be accurate. For example, if a 100Wh (10v, 10Ah) is discharged for 0.5h at 100watts, the capacity remaining should be about 50%. If that same battery is discharged at 500 watts for 0.1h, the capacity should be BELOW 50% because the "smart" battery should know that rapid discharges are less energy efficient. I think this is what "smart" batteries should know - to report energy capacity accurately.

Sorry for the long discussion... I just think it's important that this is done right. I also seek a little clarification on if Capacity Remaining remaining is always 0-100, or if it is 0-Capacity_Full.

**Comment by apache405** [comment]
> 3. Is "10" enough cells?

>No - I would go at least to 12. 6s batteries in series is not too difficult or even uncommon. 16 should cover everything in the future as well in my opinion.

I'd like to see at least 15S (http://www.ti.com/product/BQ76940) or higher.   Assuming I'm reading things correctly @hamishwillee already made it 16 (which a bq34Z100-G1 tops out at, so that is convinent).  If possible, could the number of cells be done as a `uint8_t`?  This would allow very high cell count systems to exsist.

> * Re minimum voltage per cell - is it worth adding another 16 `uint16_t` or can these all be assumed to

@hamishwillee there are two or three minimum cell voltage settings; discharge, charging, and maybe a resting value (I'd have to check the TI manuals to confirm), but they are global to all cells in the smart battery.  Same with the cell max voltages.

> Mixed chemistry batteries (individual low voltage values per cell) is something that I don't think is practical or currently available.

Agreed.  

> I also seek a little clarification on if Capacity Remaining remaining is always 0-100, or if it is 0-Capacity_Full.

@Naterater there are two state of charge estimators in the TI fuel gauges--absolute and relative.  IIRC, they both report as 0-100%, but the RSoC value is really more 0-capacity_full given the current condition of the cells and the absolute is 0-design_capacity_full.

**Comment by hamishwillee** [comment]
I'll have a think about this tomorrow, but yes, it makes sense to have two messages - one for static information and one for dynamic information that can be sent at different rates/on request (I assume this is what is meant by "static fields".

> ... If possible, could the number of cells be done as a uint8_t? This would allow very high cell count systems to exist.

@apache405 Are you proposing that we have something like:
```xml
<field type="uint16_t[15]" name="voltages" units="mV">Individual cell voltages. The cell number is the index + value of index_offset_voltages. </field>
<field type="uint8_t" name="cell_offset">The cell number of the first index in the voltages array field. Using this field allows you to specify cell voltages for batteries with more than 16 cells.</field>
```

If we reasonably expect batteries with many cells, then the extra byte is probably relatively cheap. Do we? It could be added as an extension later if needed.

What do others think?

Note, I dropped array to 15 items assuming we need 16 and they are zero indexed.

**Comment by hamishwillee** [comment]
Is cycle count static? I thought it was "running" cycle count?

**Comment by AlexKlimaj** [comment]
Cycle Count will be static per use of the battery. It will only increment on the fuel gauge once the battery is charged up after a discharge. In the case of a vehicle with autonomous battery charging, then it would increment while it is being charged.

**Comment by apache405** [comment]
@hamishwillee at a quick glance that looks like what I'm after.  The main application I can see very high cell count packs is ground vehicles.

**Comment by hamishwillee** [comment]
@apache405 Thanks. I (just) added minimum voltages including "resting" up above. Please confirm we need that/those.

> Cycle Count will be static per use of the battery. It will only increment on the fuel gauge once the battery is charged up after a discharge. In the case of a vehicle with autonomous battery charging, then it would increment while it is being charged.

Thanks. So not quite "static" (I'd been thinking truly fixed values). I changed above to static/infrequent update.

> @Naterater there are two state of charge estimators in the TI fuel gauges--absolute and relative. IIRC, they both report as 0-100%, but the RSoC value is really more 0-capacity_full given the current condition of the cells and the absolute is 0-design_capacity_full.

Do we need fields for both of those?

> @hamishwillee at a quick glance that looks like what I'm after. The main application I can see very high cell count packs is ground vehicles.

Thanks. I can see the user case. However as we go by consensus and this costs a byte, @MaEtUgR @AlexKlimaj @Naterater - thoughts?

What do people generally think about "time remaining". I agree that it is something that could be calculated externally, but what will do a better estimate than the autopilot? Should we really push this to every GCS?

**Comment by hamishwillee** [comment]
@Naterater Could you please respond to @thirtytwobits above (https://github.com/mavlink/mavlink/issues/988#issuecomment-448474272), as I am the biggest battery luddite here. 

My slightly more general question covers the same topic:

> What do people generally think about "time remaining". I agree that it is something that could be calculated externally, but what will do a better estimate than the autopilot? Should we really push this to every GCS?

Personally I'd want this for systems that might need the info but would be unlikely to calculate it - ie my Taranis RC controller. The autopilot would seem a good system to do the calculation because everything connects to it.

**Comment by Naterater** [comment]
I like how the time remaining is as proposed in the @hamishwillee post above, and I agree that it may be useful if calculated by the autopilot. https://github.com/mavlink/mavlink/issues/988#issuecomment-446017379. I envision the autopilot using an energy-based current/voltage monitor calculation instead of only a current-based calculation. Measure energy consumed, calculate or read the energy capacity of the battery, factor current time in the air (or moving), spit out estimated time remaining.

> If it isn't hard then sorry, never-mind.

It's definitely NOT easy. There are 2 schools of thought regarding representing battery capacity
1. Report the mAh consumed. Trouble is that the first 50% of mAh comes at a higher voltage than the second 50%; therefore, at 50% the battery has expended more than 50% of its energy.
2. Use the discharge curve of the battery to compensate for the effects above or measure the energy consumed (NOT JUST mAh!!). This is what I consider a "smart" battery that can accurately report its energy capacity remaining or at least how to calculate it. I think this info should probably be included in "SpecificationInfo" somehow. I also hope that message has minimum discharge voltage per cell and watt-hour rating in that message, because those values are critical IMO.

> @Naterater there are two state of charge estimators in the TI fuel gauges--absolute and relative. IIRC, they both report as 0-100%, but the RSoC value is really more 0-capacity_full given the current condition of the cells and the absolute is 0-design_capacity_full.

I don't care what is implemented, so long as it is clearly defined. Both seems redundant to me because 1 is a function of the other. Maybe we could attack it with crafted words like @hamishwillee did in his edited post above... "Charge Level" is dynamic and always 0-100% (relative), while "Capacity" represents a static value (capacity_full given the current condition of the cells).

Maybe someone else has clearer terminology or thoughts. I really only have a grasp on the concepts of smart batteries -- I'm not a guru.

**Comment by hamishwillee** [comment]
Thanks @Naterater .

1. I updated `MAV_SMART_BATTERY_FAULT_DELTA` to new name. Hope OK. How would someone use this? Ie I assume a single cell fail is somehow not critical? Is the assumption the capacity and voltage as coming out of the other fields would change as long as this?
   ```xml
   <entry value="4" name="MAV_SMART_BATTERY_FAULT_SINGLE_CELL_FAIL">
        <description>Single cell has failed.</description>
      </entry>
   ```

2. Re charge estimators

>> @Naterater there are two state of charge estimators in the TI fuel gauges--absolute and relative. IIRC, they both report as 0-100%, but the RSoC value is really more 0-capacity_full given the current condition of the cells and the absolute is 0-design_capacity_full.

> I don't care what is implemented, so long as it is clearly defined. Both seems redundant to me because 1 is a function of the other. Maybe we could attack it with crafted words like @hamishwillee did in his edited post above... "Charge Level" is dynamic and always 0-100% (relative), while "Capacity" represents a static value (capacity_full given the current condition of the cells).

OK, so we currently have for this:
```
      <field type="int8_t" name="capacity_remaining" units="%">Remaining battery energy. Values: [0-100], -1: field not provided.</field>
```
My first reaction was that we should use the "relative approach" because our `capacity_full` takes account of battery condition. So we would change above to:
```
      <field type="int8_t" name="capacity_remaining" units="%">Remaining battery energy (as a percentage of capacity_full). Values: [0-100], -1: field not provided.</field>
```
Another alternative is to change this to an absolute capacity in mAh. Percentage is better if this is just for presentation in GCS. I guess absolute capacity better if consumers want to use it for calculations of time remaining or whatever - but I don't know if the estimate would ever be good enough that this is even an option.

Thoughts?

**Comment by hamishwillee** [comment]
No one commented on the high cell count option, but I added it anyway :-) I figure that with a default value of zero and at the end of the message it will be trimmed in most cases.

For "time remaining" there is one comment in support and one "not that useful". As feelings are ambivalent, I am leaving it in. 

Do we need all the "dynamic" messages at similar rate? ie should we perhaps split out cell voltages into their own message?

**Comment by MaEtUgR** [comment]
> dropped array to 15 items assuming we need 16 and they are zero indexed.

@hamishwillee In C syntax an `uint16_t array[16];` has 16 elements with indexes 0, 1, ..., 15. Is that different here?

> Is cycle count static?

@hamishwillee Presumably static over one flight (except when charged in flight). Sorry, I should have been more specific. My goal was to prevent high-frequency streaming of nearly constant data. E.g. cell voltage and battery remaining should typically be updated in real-time and therefore be streamed with maybe ~0.5-10Hz while weight should in theory not change at all and all and cycle count can be updated by the reporting component whenever there is a change. So the solution in my eyes like you started is to separate these types and have one message streamed while the other is sent once at the beginning and can be updated whenever necessary because there's an "unexpected" change.

> `<field type="int16_t" name="current" units="cA">Battery current (through all cells/loads). Positive if discharging, negative if charging. NaN: field not provided.</field>`

@hamishwillee Only floats can be set to NaN.

> If we reasonably expect batteries with many cells, then the extra byte is probably relatively cheap.

If you build a big drone with a lot of parallel batteries and it reports the smart battery status for some reason with one message for the whole drone. Then you can easily go over 16 cells. Look at the DJI M600 with a 6 cell smart battery per arm resulting in 36 total cells when fully equiped. The byte seems worth spending. I would argue that you could save a lot better by decreasing the array size to common battery sizes like e.g. 6 cells and using the `cell_offset` as soon as the simple default sizes are exceeded. Is there a padding reson why `voltages[]` should not be just before `cell_offset`?

> <field type="uint16_t" name="charging_minimum_voltage" units="mV">Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.</field>

Sorry if I got the wires crossed, what is this used for?
I had the idea to specify the batteries chemistry/type e.g. LiPo, LiHv, ... to provide via fixed specs more details about its properties like maximum and minimum cell voltage. According to my experience the rough value minimum voltage is given by the battery type and exact values are more like a religion depending on priorities like flight time vs. battery life durability given how you store and balance them.

> What do people generally think about "time remaining". I agree that it is something that could be calculated externally, but what will do a better estimate than the autopilot? Should we really push this to every GCS?

We need to be clear about what this message is for. I presumed it's mainly for a smart battery controller supporting MAVLink to report to the autopilot or other components including GCS interested in **raw information**. According to my plans the autopilot would then provide the processed information like the "definitely NOT easy" to calculate "time remaining". If there was no autopilot to take that role it would make sense to have an inaccurate time remaining estimate from the smart battery instead of none at all but the one from the autopilot is worst case the same and best case a lot more accurate. In any case the GCS should not have to calculate anything complicated from these raw values. It should either present them because they are interesting or have other messages/fields to rely on processed information. Please correct me if we speak about different things here.

> There are 2 schools of thought regarding representing battery capacity

@Naterater Interesting thought! Having an accurate mAh measurement with a persistant state over vehicle reboots and charging cycles is already a huge step compared to a "dumb battery" but you're making a good point here. It seems beneficial at least from a theoretical standpoint to accumulate energy (Wh) instead of electric charge (Ah) because the voltage is expected to fluctuate in the order of 20% for normal LiPo usage. If for some reason the battery is not capable of any accurate reports or you want to use the energy to calculate the remaining time in a simple way the calculations can still be approximated by assuming constant nominal voltage. I'm open to aynone having practical experience with using the energy based approach.

**Comment by hamishwillee** [comment]
@MaEtUgR Here are some responses. I'm going to split them into a couple of posts. 

>> dropped array to 15 items assuming we need 16 and they are zero indexed.

> @hamishwillee In C syntax an uint16_t array[16]; has 16 elements with indexes 0, 1, ..., 15. Is that different here?

No, I am an idiot. Fixed above. 

>> Presumably static over one flight (except when charged in flight). 

> Sorry, I should have been more specific. My goal was to prevent high-frequency streaming of nearly constant data. E.g. cell voltage and battery remaining should typically be updated in real-time and therefore be streamed with maybe ~0.5-10Hz while weight should in theory not change at all and all and cycle count can be updated by the reporting component whenever there is a change. So the solution in my eyes like you started is to separate these types and have one message streamed while the other is sent once at the beginning and can be updated whenever necessary because there's an "unexpected" change.

Yes, got that! You said "like you started is to separate these types". I thought I'd finished separating them (ie everything in the [SMART_BATTERY_STATUS](https://github.com/mavlink/mavlink/issues/988#issuecomment-447699375) should be streamed, and the rest only needs to be sent once). Did I miss one?

>> <field type="int16_t" name="current" units="cA">Battery current (through all cells/loads). Positive if discharging, negative if charging. NaN: field not provided.</field>

> Only floats can be set to NaN.

Fixed above - changed to UINT16_MAX

**Comment by hamishwillee** [comment]
@MaEtUgR 

>> If we reasonably expect batteries with many cells, then the extra byte is probably relatively cheap.

> If you build a big drone with a lot of parallel batteries and it reports the smart battery status for some reason with one message for the whole drone. Then you can easily go over 16 cells. Look at the DJI M600 with a 6 cell smart battery per arm resulting in 36 total cells when fully equipped. The byte seems worth spending. 

OK, so worth the byte. Good, since I already added it :-)

> Is there a padding reason why `voltages[]` should not be just before `cell_offset`?

No, I have moved it. 
Ordering matters in some cases but not this one - smaller data types are ordered last.

> I would argue that you could save a lot better by decreasing the array size to common battery sizes like e.g. 6 cells and using the `cell_offset` as soon as the simple default sizes are exceeded. 

What if you have a 36 cell battery? Doesn't that mean we have to potentially  send the temperature (and other data that doesn't update much) 6 times "unnecessarily"?

I'm not raising this to argue the logic - but to suggest perhaps we could instead spend 1 more message on this just for voltages. The nature of 

```xml
    <message id="366" name="SMART_BATTERY_CELL_VOLTAGES">
      <description>Smart Battery cell voltage information.</description>
      <field type="uint16_t[255]" name="voltages" units="mV">Individual cell voltages.  Batteries with more 16 cells can use the cell_offset field to specify the cell offset for the array specified in the current message . Index values above the valid cell count for this battery should have the 0 value.</field>
    </message>
```

Note:
- No offset - though we can add one if you think we'd need more than 254 cells. That should be a uint16_t to allow it to be ordered in declaration order so we can truncate unused cells.
- Default value of 0, which I assume is unlikely for a real cell. The reason is that this allows any empty cells to be truncated for sending - which is why this is so efficient - equally good for 6 cells or 254.

**Comment by hamishwillee** [comment]
>> <field type="uint16_t" name="charging_minimum_voltage" units="mV">Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.</field>

> Sorry if I got the wires crossed, what is this used for?
>I had the idea to specify the batteries chemistry/type e.g. LiPo, LiHv, ... to provide via fixed specs more details about its properties like maximum and minimum cell voltage. According to my experience the rough value minimum voltage is given by the battery type and exact values are more like a religion depending on priorities like flight time vs. battery life durability given how you store and balance them.

@Naterater @apache405 I think that question for you!

**Comment by hamishwillee** [comment]
>> What do people generally think about "time remaining". I agree that it is something that could be calculated externally, but what will do a better estimate than the autopilot? Should we really push this to every GCS?

> We need to be clear about what this message is for. I presumed it's mainly for a smart battery controller supporting MAVLink to report to the autopilot or other components including GCS interested in **raw information**. According to my plans the autopilot would then provide the processed information like the "definitely NOT easy" to calculate "time remaining". If there was no autopilot to take that role it would make sense to have an inaccurate time remaining estimate from the smart battery instead of none at all but the one from the autopilot is worst case the same and best case a lot more accurate. In any case the GCS should not have to calculate anything complicated from these raw values. It should either present them because they are interesting or have other messages/fields to rely on processed information. Please correct me if we speak about different things here.

Yes, that's the sort of thing I was thinking. My main use case is actually knowing how much flight time I might have left from my Taranis. I was hoping "something" could calculate this, and if a system doesn't provide the information that is fine too. 

Let's vote on this field.  Thumbs up to keep, Thumbs down to remove.

**Comment by Naterater** [comment]
> > > Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.
> 
> > Sorry if I got the wires crossed, what is this used for?
> > I had the idea to specify the batteries chemistry/type e.g. LiPo, LiHv, ... to provide via fixed specs more details about its properties like maximum and minimum cell voltage. According to my experience the rough value minimum voltage is given by the battery type and exact values are more like a religion depending on priorities like flight time vs. battery life durability given how you store and balance them.
> 
> @Naterater @apache405 I think that question for you!

This should not be a minimum when **charging** - it should be a minimum when **discharging** (ie. a recommended minimum voltage under load). For some li-ion cells, this is as low as 2.5v. For most li-po cells, it is 3.5v or more. I think this is important to leave up to the smart battery manufacturers.

There should also be a maximum voltage value - typically 4.2v, however Li-Hv cells go up to 4.35v. Useful for recharging batteries with different chemistry.

Failsafe voltages can always be set by the autopilot, so "religions" on these voltages can always be overridden by religious users. For the rest of us, the safe voltages can be given by the battery itself, and we will let the AP decide when to return home when the voltages are low enough compared to these values. I kind of get the idea to do something like an enum to specify chemistry and have a library of voltages for specific chemistry, however I think this battery technology is ever changing, and maintaining a library with all of the new battery types would be a nightmare.

**Comment by hamishwillee** [comment]
@Naterater 

> This should not be a minimum when charging - it should be a minimum when discharging (ie. a recommended minimum voltage under load)

You proposed three setpoints, which I included in the proposed message. I THINK that `discharge_minimum_voltage` below is correct, but that `charging_minimum_voltage` is incorrect - it should be charging **maximum** voltage right? Are the others OK?
```
      <field type="uint16_t" name="discharge_minimum_voltage" units="mV">Minimum per-cell voltage when discharging. If not supplied set to UINT16_MAX value.</field>
      <field type="uint16_t" name="charging_minimum_voltage" units="mV">Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.</field>
      <field type="uint16_t" name="resting_minimum_voltage" units="mV">Minimum per-cell voltage when resting. If not supplied set to UINT16_MAX value.</field>
```

**Comment by LorenzMeier** [comment]
@hamishwillee I think we need to discuss on one of the next calls and go through it.

**Comment by hamishwillee** [comment]
I'm going to create a PR with this to move it a step closer to reality as discussed with Lorenz in the last meeting. 

One question. We have `capacity_remaining` (%), which I guess can be used in the same way as [MAV_BATTERY_CHARGE_STATE](https://github.com/mavlink/mavlink/blob/25814dd7ac5283f71743be8108e0fce86a4a9963/message_definitions/v1.0/common.xml#L2605) except that there is no easy way to detect if it is charging or not. We might need this information for the reasons given in https://github.com/mavlink/mavlink/pull/1068

Thoughts?

**Comment by hamishwillee** [comment]
@AlexKlimaj Thanks. As defined we're doing this for current - but with charging defined as a negative value.

@MaEtUgR I understand your answer, but I'm not sure if you're using it as an argument for or against adding charge state information as an enum!
I think that you're probably saying no - we provide information to the GCS about state, and it decides whether it wants to report this as critical or not. Correct?

**Comment by MaEtUgR** [comment]
@hamishwillee Sorry I lost you there. The autopilot should report a battery state enum to components that want to stay informed (mostly the ground station) that's what my explanation was for. IMHO the smart battery doesn't need to report such a charge state to autopilot or ground station. If it's being charged can be seen from the negative current like @AlexKlimaj mentioned and the conclusions about the current situation are drawn by the autopilot and reflected in the battery status it sends out. Did I by any chance cover your questions?

**Comment by hamishwillee** [comment]
Thanks. Question answered!

**Comment by hamishwillee** [comment]
Concrete smart battery proposal now here: https://github.com/mavlink/mavlink/pull/1072

Please address any final thoughts on that issue. I'm going to close this now.
